### PR TITLE
Improve error message when mounting non-string/function elements

### PR DESCRIPTION
**Pull request by spicyj**
People keep getting tripped up by the nonsense error you currently get and manage not to see the warning. See #3478. I think this check is cheap enough. (We also have a warning at element creation time, which is more useful if you notice it.)

### CODE REVIEW

File: src/renderers/shared/reconciler/__tests__/ReactComponent-test.js [hunk 1]
```diff
@@ -279,4 +279,32 @@ describe('ReactComponent', function() {
...
+      'built-in components) or a class/function (for composite components) ' +
+      'but got: null.'
+    );
+
+    var Z = {};
+    expect(() => ReactTestUtils.renderIntoDocument(<Z />)).toThrow(
+      'Invariant Violation: Element type is invalid: expected a string (for ' +
+      'built-in components) or a class/function (for composite components) ' +
+      'but got: object.'
+    );
```
Comment (zpao) [inline src/renderers/shared/reconciler/__tests__/ReactComponent-test.js]: Want to add one of these testamajiggers for the addendum?

Comment (spicyj) [inline src/renderers/shared/reconciler/__tests__/ReactComponent-test.js]: I have a secret test already in src/isomorphic/classic/element/__tests__/ReactElementValidator-test.js…

Comment (zpao) [inline src/renderers/shared/reconciler/__tests__/ReactComponent-test.js]: so you do

**Comment by infolock** [comment]
:+1:

**Pull request merged**